### PR TITLE
fix: Stuck at "Retrieving character list".

### LINKED ISSUE #877: Stuck at "Retrieving character list" for 5 minutes
**Description**:

Irregularly, but often enough, when we launch AzerothCore we are stuck at "retrieving character list" just after login on our account. This concerns all the people logging in (players).

**Current behaviour**:

We are stuck during 2 minutes to 5 minutes (maybe more, sometimes we are stuck for more and it times out - this is not verified). Once we are able to choose our characters and enter the game, it never happens again (see **Remarks** below).

**Expected behaviour**:

This should never happen

**Remarks**:

- This has been noticed by MANY people in the discord, by me personally, by AyaseCore too.
- @AyaseCore told me this didn't occur on SunwellCore. This need to be verified (same for mangos, cmangos, trinitycore etc)
- It happened to me only when I didn't start my server for months. Or switched to another unused for a while database (like an older characters db). **So I think this is related to some MySQL cache or something like that** - need opinions on this.

**Steps to reproduce the problem**:

I don't know, it seems like people here have a similar issue (not sure if it's the same) https://github.com/azerothcore/azerothcore-wotlk/issues/711#issuecomment-362535892

**Branch(es)**: 0.x / 1.x / master (Specify the branch(es) affected by this issue) 

**AC hash/commit**:  

**Operating system**:  
On windows or on linux, same thing

PS: I started writing this issue and noticed the other issue afterwards so I let this one to refresh the problem https://github.com/azerothcore/azerothcore-wotlk/issues/711#issuecomment-362535892

<bountysource-plugin>

---
Want to back this issue? **[Post a bounty on it!](https://www.bountysource.com/issues/58465793-stuck-at-retrieving-character-list-for-5-minutes?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github)** We accept bounties via [Bountysource](https://www.bountysource.com/?utm_campaign=plugin&utm_content=tracker%2F40032087&utm_medium=issues&utm_source=github).
</bountysource-plugin>

### PR DESCRIPTION
**Pull request by AyaseCore**
**Changes proposed:**

**Target branch(es):**  master

**Issues addressed:**  Closes #877 #711 

**Tests performed:** tested in-game

**Known issues and TODO list:** none

**NOTE** You no longer need to squash your commits, on merge we will squash it for you. (GitHub added a new feature)

------

If you need to verify it.
Before applying this fix.Running the SQL script in characters DB:
```
update instance_reset set resettime = 12345;
```
Restart worldserver.Login account. You will find stuck at "Retrieving character list".
Apply this fix. Repeat the above steps. You can enter the game.

### CODE REVIEW

**Comment by AyaseCore** [comment]
In fact, I think this issue is rare.
Only worldserver is a long time not run.This issue will arise.
It needs to update resttime long ago to the present time.
So we have many solutions.

1. Clean up too old data before reading the instance_reset table.
2.Outside the loop. Execute the SQL command.
3.Increase judgment. Reduce SQL command execution.
4.Modified algorithm. For too old data.So let's calculate the final number.Then execute the SQL command.

**Comment by BarbzYHOOL** [comment]
Ok I tried this ```update instance_reset set resettime = 12345;``` on the character DB, and when you connect, you can look at instance_reset and it keeps updating little by little (refresh the view each time).

Of course it takes so long that my server crashes because of the anti stuck configuration "Starting up anti-freeze thread (30 seconds max stuck time)..."

**Comment by Stoabrogga** [comment]
Could someone please explain to me why this should work? Those are all integer values:
next_reset
period
diff
MINUTE (60)
DAY (86400)

Division by integer has no fractional part, e.g. for initial value of 1527088446 for next_reset:
1527088446 + MINUTE = 1527088506
1527088506 / DAY = 17674 (no fractional part)
17674 * DAY = 1527033600

So the calculation

((next_reset + MINUTE) / DAY * DAY)

will always lead to the same value because adding a minute doesn't change the first division (would only make a difference if working with floats). Or am I totally mislead in this case?

**Comment by AyaseCore** [comment]
I guess:
if next_reset = 1527119962(2018/5/24 7:59:22)
1527119962/ DAY = 17674
17674 * DAY = 1527033600(2018/5/23 8:0:0)

1527119962+ MINUTE = 1527120022(2018/5/24 8:0:22)
1527120022/ DAY = 17675
17675 * DAY = 1527120000(2018/5/24 8:0:0)

**Comment by Stoabrogga** [comment]
Sure, but if you have another value for next_reset which lies in the past and is not exactly at the edge of another day you're stuck in the while loop because the value stays always the same.

**Comment by Rochet2** [comment]
> the calculation
> ((next_reset + MINUTE) / DAY * DAY)
> will always lead to the same value because adding a minute doesn't change the first division

The +MINUTE does shift the time calculation a bit in the edge case where time is near the end of a day as ayase noted, but why that would be needed is not clear to me either.
Further inspection shows that the code is 10 years old and can only be traced back to the SVN -> git transition back in 2008.
https://github.com/TrinityCore/TrinityCore/commit/9b1c0e006f20091f28f3f468cfcab1feb51286bd

> you're stuck in the while loop because the value stays always the same

Only if period and diff would be 0. Since they should not be, the time will increase by some amount of seconds and will not be infinite. But it seems that indeed +MINUTE alone does not increase time in pretty much all cases, if that was your point.



I am not familiar with how raid resets are supposed to work. In the code I posted I simply copied the existing logic, but it partially ignores logic in InstanceSaveManager::Update where `bool warn = event.type < 5;` is true. I guess that should be added in at least to keep consistency with the reset times, I cant see if there is other relevant logic I would have missed.

**Comment by BarbzYHOOL** [comment]
I have people who can testify that I wanted solution 4!! :rofl: 

I tested and effectively, it only modified each row once (so only one query per row, only one calculation I guess) and freezing was very short.

Thank you @AyaseCore and @Rochet2 

If there is something else to add or edit, tell me before we merge it

**Comment by Stoabrogga** [comment]
@Rochet2 I think I've got it, thanks. It will only work if the sum of period and diff is greater than or equal to 1 day, which should always be the case.

Hmm, you could also calculate the value directly instead of the while loop. This here should also work:

```
if (time_t(next_reset - HOUR) < now)
{
    uint32 skip_minutes = uint32((now - resetTime + HOUR + MINUTE) / MINUTE);
    next_reset = uint32(((next_reset + (skip_minutes * MINUTE)) / DAY * DAY) + period + diff);
}
```

**Comment by BarbzYHOOL** [comment]
Just retested and it's longer than 40ms because I could refresh 3 times before all the rows were filled :P

**Comment by AyaseCore** [comment]
Either solution 3 or solution 4.
The goal is to reduce unnecessary SQL command execution.
Personally, I prefer solution 3, because it may only take one line of code to solve the issue. :grinning:

**Comment by BarbzYHOOL** [comment]
@Rochet2 said this can be merged. I tested and it's working very well.

Thanks a lot to Rochet2 and @AyaseCore for starting this!

File: src/server/game/Instances/InstanceSaveMgr.cpp [hunk 1]
```diff
@@ -290,6 +290,12 @@ void InstanceSaveManager::LoadResetTimes()
             SetResetTimeFor(mapid, difficulty, t);
             CharacterDatabase.DirectPExecute("INSERT INTO instance_reset VALUES ('%u', '%u', '%u')", mapid, difficulty, (uint32)t);
         }
+        else
+        {
+            // next reset should be in future. If its not, skip to future.
+            while (t < now)
+                t = uint32(((t + MINUTE) / DAY * DAY) + period + diff);
```
Comment (ShinDarth) [inline src/server/game/Instances/InstanceSaveMgr.cpp]: is `DAY * DAY` meant to do some rounding?

Comment (Stoabrogga) [reply src/server/game/Instances/InstanceSaveMgr.cpp]: Not the `DAY * DAY` but the `/ DAY * DAY`. The first division by DAY leads to the number of days without fractional part. Multiplicating with DAY results in the exact number of seconds for x days.

Comment (ShinDarth) [reply src/server/game/Instances/InstanceSaveMgr.cpp]: yeah that's what I meant, thanks for clarifying my doubt!

**Comment by BarbzYHOOL** [comment]
Better fix here https://github.com/azerothcore/azerothcore-wotlk/commit/c4a6445a232a3cd8fe8d1e636f91b98c68900943

**Pull request merged**